### ISSUE TITLE
Map freeze if easeTo during move event

### ISSUE THREAD

**Issue by ToHold** [maplibre/maplibre-gl-js#4010]
**maplibre-gl-js version**: 4.1.2

**browser**: Any

### Expected Behavior
I am using maplibre as map tiled core for a 3D project. I'm also using threejs to handle my own 3D elements. In my case I would like to automatically change map pitch when I zoom uppon level 14. Whether with NavigationControl, scrollZoom, or touchZoomRotate.

### How I do the trick
I listen on "zoom" event, that is fired when any movement that change map zoom is done. Inside this listener I store in a named var lastZoom the current zoom, then I check in the next event if my lastZoom <= 14 && zoom >= 14, if yes i change the pitch using an easeTo pitch 40, otherwise if  lastZoom >= 14 && zoom <= 14 I easeTo a 0 pitch.

```
let lastZoom = map.getZoom();
map.on('zoom', () => {
	const zoom = map.getZoom();
	if (lastZoom <= 14 && zoom >= 14) {
				
		map.easeTo({pitch: 40}, { customEase: true});
				
	} else if(lastZoom >= 14 && zoom <= 14) {
				
		map.easeTo({pitch: 0, bearing: 0}, { customEase: true});
				
	}
			
	lastZoom = zoom;
});
```

### Actual Behavior
In some case, if I move, or I keep zooming with NavigationControl the map would freeze.
I tried to investigate, and I found that the problem occurs because Camera._stop is called during a Camera._onEaseFrame. That should delete Camera._easeFrameId, but because we are running an easeTo inside the "zoom" event, it also creates a new Camera._easeFrameId while the algorithm was not intended for. And then it cause an Camera._easeFrameId offset that also cause a freeze.

### Link to Demonstration
You just have to triple fast click on zoomIn button, or to try moving by drag and drop during the ease.
https://codepen.io/ToHold/pen/oNOBgyg

### Pull request
I proposed a working solution here : https://github.com/maplibre/maplibre-gl-js/pull/3853
But it was confusing because I used an Array to handle the multiple _easeFrameId, and indeed, if someone read the code and try to improve it, it can be harder to understand what was intended for.

### Help
I would apreciate to get help to achieve the desired behavior whithout any freeze case.

**Comment by ToHold** [comment]
It seems to work fine for zoomIn zoomOut : https://codepen.io/ToHold/pen/XWQPzXG
But if you try to scroll, map center is changing, any idea why ? I guess it can come from the fact that scroll_zoom handler is also using transform at the same time.

**Comment by HarelM** [comment]
I don't think you should use easeTo in that method, but rather return the updated state, see here:
https://maplibre.org/maplibre-gl-js/docs/API/types/CameraUpdateTransformFunction/

**Comment by ToHold** [comment]
Yes I thought about it, but because it's a "before update callback" I don't think that I'll be able to ease anyway, even if I create my own easing function using _requestRenderFrame

**Comment by HarelM** [comment]
I've written the following callback, it's not perfect, but I think one can tweak it a bit more to receive the needed effect:
```
map.transformCameraUpdate = (transform) => {
  let copy = {...transform};
  if (transform.zoom > 13.5 && transform.zoom < 14.5 && !copy.pitch) {
    copy.pitch = 60 * (transform.zoom - 13.5);
  } else if (transform.zoom <= 13.5 && !copy.pitch) {
    copy.pitch = 0
  } else if (transform.zoom >= 14.5 && !copy.pitch) {
    copy.pitch = 60
  }
  return copy;
};
```

**Comment by ToHold** [comment]
Thank's I think that's a great work arround ! I'll try to implement it in my case.

**Comment by ToHold** [comment]
I managed to get the desired behavior thanks to you, thank you. However, I was forced to stop the animations when transform.zoom >= 14.5 and when transform.zoom <= 13.5 because it could lead to weird behavior. 

```
updateCameraPitch(transform: {
		center?: MapLibreGL.LngLat;
		zoom?: number;
		pitch?: number;
		_pitch?: number;
		bearing?: number;
		elevation?: number;
	}) {

		try {
			const step = 0.5;
			const zoom3dThreshold = this._initialOpts.zoom3dThreshold - step;

			if (transform.zoom > zoom3dThreshold - step && transform.zoom < zoom3dThreshold + step) {

				transform.pitch = DefaultPitch * (transform.zoom - (zoom3dThreshold - step));
				this._needSetDefaultPitch._3d = true;
				this._needSetDefaultPitch._2d = true;

			} else if (transform.zoom <= zoom3dThreshold - step && this._needSetDefaultPitch._2d) {
				
				this._needSetDefaultPitch._2d = false;
				
				// without stop, an already started camera movement, can overide map transform
				// such as scroll handler or zoom in zoom out buttons
				this._map.stop();
				transform.pitch = 0;
				transform.bearing = 0;

				this._needSetDefaultPitch._3d = true;

			} else if (transform.zoom >= zoom3dThreshold + step && this._needSetDefaultPitch._3d) {

				this._needSetDefaultPitch._3d = false;

				// without stop, an already started camera movement, can overide map transform
				// such as scroll handler or zoom in zoom out buttons
				this._map.stop();
				transform.pitch = DefaultPitch;

				this._needSetDefaultPitch._2d = true;

			}
	
			return transform;
		} catch (error) {
			console.error(error);
			return transform;
		}
	}
```
